# Python_DataManipulation_EDA_1

In [3]:
# import libraries
import pandas as pd
import numpy as np
%matplotlib in line

# set maximum columns display as 100
pd.set_option('display.max_columns',100) 

UsageError: unrecognized arguments: line


In [4]:
# import the dataset
df = pd.read_csv('Salaries.csv', )
df.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
1  2011    NaN  San Francisco     NaN  
2  2011    NaN  San Francisco     NaN  
3  2011    NaN  San Francisco     NaN  
4  2011    NaN  San Francisco     NaN

### 1. Check the size, columns and data types

In [5]:
# check data size
print(df.shape) # returns a tuple
print(type(df.shape))

# check column names
print(df.columns)

# check data types in each column
print(df.dtypes)

(148654, 13)
<class 'tuple'>
Index(['Id', 'EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year', 'Notes', 'Agency',
       'Status'],
      dtype='object')
Id                    int64
EmployeeName         object
JobTitle             object
BasePay             float64
OvertimePay         float64
OtherPay            float64
Benefits            float64
TotalPay            float64
TotalPayBenefits    float64
Year                  int64
Notes               float64
Agency               object
Status              float64
dtype: object


### 2. Check missing or non-missing values 

#### missing values for the whole dataframe

In [6]:
# generate a summary of missing values
print(df.isnull().sum())# missing values

print(df.notnull().sum()) # not missing values

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                609
OvertimePay              4
OtherPay                 4
Benefits             36163
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              148654
dtype: int64
Id                  148654
EmployeeName        148654
JobTitle            148654
BasePay             148045
OvertimePay         148650
OtherPay            148650
Benefits            112491
TotalPay            148654
TotalPayBenefits    148654
Year                148654
Notes                    0
Agency              148654
Status                   0
dtype: int64


In [7]:
# check overall whether a column in df has missing values
df.isnull().any() # if missing - True, not missing - False
# returns a pd series

Id                  False
EmployeeName        False
JobTitle            False
BasePay              True
OvertimePay          True
OtherPay             True
Benefits             True
TotalPay            False
TotalPayBenefits    False
Year                False
Notes                True
Agency              False
Status               True
dtype: bool

In [8]:
df.notnull().any() # if missing - False, not missing - True
# returns a pd series

Id                   True
EmployeeName         True
JobTitle             True
BasePay              True
OvertimePay          True
OtherPay             True
Benefits             True
TotalPay             True
TotalPayBenefits     True
Year                 True
Notes               False
Agency               True
Status              False
dtype: bool

In [9]:
type(df.notnull().any())

pandas.core.series.Series

In [10]:
# list out all columns with missing values
df.columns[df.isnull().any()]

Index(['BasePay', 'OvertimePay', 'OtherPay', 'Benefits', 'Notes', 'Status'], dtype='object')

In [11]:
# list out all columns without any missing values
df.columns[df.notnull().any()].tolist() 

['Id',
 'EmployeeName',
 'JobTitle',
 'BasePay',
 'OvertimePay',
 'OtherPay',
 'Benefits',
 'TotalPay',
 'TotalPayBenefits',
 'Year',
 'Agency']

In [12]:
# use list comprehension
print('Columns WITH missing values', [col for col in df.columns if df[col].isnull().any()])
print('Columns WITHOUT missing values', [col for col in df.columns if df[col].notnull().any()])

Columns WITH missing values ['BasePay', 'OvertimePay', 'OtherPay', 'Benefits', 'Notes', 'Status']
Columns WITHOUT missing values ['Id', 'EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay', 'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year', 'Agency']


In [13]:
# or use 
print(df.loc[:, df.isnull().any()].columns)
print(df.loc[:, df.isnull().any()].columns)

Index(['BasePay', 'OvertimePay', 'OtherPay', 'Benefits', 'Notes', 'Status'], dtype='object')
Index(['BasePay', 'OvertimePay', 'OtherPay', 'Benefits', 'Notes', 'Status'], dtype='object')


In [14]:
# check detail numbeer of columns with missing values
print(df.isnull().any().value_counts()) 
print('There are in total', df.isnull().any().value_counts()[1], 'with missing values')
print('There are in total', df.isnull().any().value_counts()[0], 'without missing values')

False    7
True     6
dtype: int64
There are in total 6 with missing values
There are in total 7 without missing values


The data set has 148,654 rows and 13 columns; 
In terms of missing values, there are 6 columns:
- 609 missing values in 'BasePay'
- 36,163 missing valies in 'Benefits'
- 4 missing values each in 'OvertimePay' and 'OtherPay'  
- there are no values assigned in 'Notes' and 'Status' columns

#### missing values for individual column

**.value_counts():** 
- for checking individual columns, not work for the whole df 

- use .isnull().value_counts() or .notnull().value_counts() will get the same results; 
- pay attention to which categoty (missing or non-missing) is true or false


In [15]:
# check total number of null values for an indivudual column
print(df['BasePay'].isnull().value_counts()) 
# or usedf.BasePay.isnull().value_counts()

print(df['BasePay'].notnull().value_counts())

False    148045
True        609
Name: BasePay, dtype: int64
True     148045
False       609
Name: BasePay, dtype: int64


In [16]:
# check the number of null values and notnull values in a columns
print(df['BasePay'].isnull().sum())
print(df['BasePay'].notnull().sum())

609
148045


In [17]:
# or use below format to check the number of null values and notnull values in a columns
print(pd.isnull(df.BasePay).sum())
print(pd.notnull(df.BasePay).sum())

609
148045


#### percentage of missing values 

**.shape returns tuple** 
- so you need to extract the size of the dataframe or size of a column
- size of a dataframe: .shape[0] to get number of rows; .shape[0] to get number of columns
- size of a column:.shape[0]; .shape[1] will not work as  df.col.shape returns a tuple like this (###, )

**or use len()**
- for the dataframe or a single column

In [18]:
print(df.shape)
print(df.shape[0])
print(df.shape[1])
print(len(df))

(148654, 13)
148654
13
148654


In [19]:
# check % of missing values
pd.isnull(df.BasePay).sum()/df.BasePay.shape[0]

0.004096761607491221

#### summary function for missing values

In [20]:
def missing_values_summary(df):
        mis_values = df.isnull().sum()
        mis_values_percent = 100 * df.isnull().sum() / df.shape[0]
        
        # create a table summary
        summary_table = pd.concat([mis_values, mis_values_percent], axis=1)
        summary_table = summary_table.rename(columns = {0 : 'Missing Values', 1 : '% of Total Values'})
        
        summary_table['Data Type'] = df.dtypes
        
        summary_table = summary_table[summary_table.iloc[:,1] != 0].sort_values('% of Total Values', ascending=False).round(4)
        
        # print out overall summary
        print("This dataframe has {df.shape[0]} Rows and {df.shape[1]} Columns.".format(df = df))  
        print("There are {} columns that have missing values.".format(summary_table.shape[0]))
        print("The columns that have missing values are: ", [col for col in df.columns if df[col].isnull().any()])
        print("The columns that do not have missing values are: ", [col for col in df.columns if df[col].notnull().any()])
        
        return summary_table

missing_values_summary(df)


# could save this table to excel
# summary_table.to_excel('missing_values_summary.xlsx', freeze_panes=(1,0), index = False)

This dataframe has 148654 Rows and 13 Columns.
There are 6 columns that have missing values.
The columns that have missing values are:  ['BasePay', 'OvertimePay', 'OtherPay', 'Benefits', 'Notes', 'Status']
The columns that do not have missing values are:  ['Id', 'EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay', 'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year', 'Agency']


Missing Values  % of Total Values Data Type
Notes                148654           100.0000   float64
Status               148654           100.0000   float64
Benefits              36163            24.3270   float64
BasePay                 609             0.4097   float64
OvertimePay               4             0.0027   float64
OtherPay                  4             0.0027   float64

In [21]:
# or use below function
def missing(df):
    print("This dataframe has {df.shape[0]} Rows and {df.shape[1]} Columns.".format(df = df))  
    print("There are {} columns that have missing values.".format(df.isnull().any().value_counts()[1]))
    return round((df.isnull().sum() * 100/ len(df)),4).sort_values(ascending=False)

missing(df)

This dataframe has 148654 Rows and 13 Columns.
There are 6 columns that have missing values.


Status              100.0000
Notes               100.0000
Benefits             24.3270
BasePay               0.4097
OtherPay              0.0027
OvertimePay           0.0027
Agency                0.0000
Year                  0.0000
TotalPayBenefits      0.0000
TotalPay              0.0000
JobTitle              0.0000
EmployeeName          0.0000
Id                    0.0000
dtype: float64

**Missing values:**
- normally compare the % with 5% (if < 5% can delete them when there is no pattern exist among these missing values)


### 3. Check unique values

In [22]:
# check unique values
print(df['Year'].unique()) # returns a numpy array of unique values

print(df['Year'].nunique())  # return the total number of unique values

print(type(df['Year'].unique()))


[2011 2012 2013 2014]
4
<class 'numpy.ndarray'>


## 4. Slicing data with .loc and .iloc
- .loc[row_num , col_name_str]: string based slicing
  - Inclusive on left and right side
  
- .iloc[row_num, col_name_int] : integer based slicing, iloc for 'integer locations'
  - Not inclusive on the right side

### use .loc[ , ] to select rows and columns

In [23]:
# choose 1 columns
df_sub_loc1 = df.loc[100:120, 'BasePay']
df_sub_loc1

# will get 20 rows for column 'BasePay' indexed from 100-120 (inclusive)

100    168692.65
101    123105.00
102    105934.69
103     88400.70
104    105934.64
105    110474.93
106    149362.86
107    188708.00
108    188708.00
109    105934.67
110    206057.69
111     63811.18
112    220115.22
113    188708.00
114     88400.70
115    121628.34
116    219804.34
117    191054.19
118    150432.85
119    179648.00
120    140546.88
Name: BasePay, dtype: float64

In [24]:
# choose all rows for multiple columns
df_sub_loc2 = df.loc[:, ['JobTitle','BasePay', 'Year']]
df_sub_loc2.head()

# will get all rows for columns: 'JobTitle','BasePay', 'Year'

JobTitle    BasePay  Year
0  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18  2011
1                 CAPTAIN III (POLICE DEPARTMENT)  155966.02  2011
2                 CAPTAIN III (POLICE DEPARTMENT)  212739.13  2011
3            WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00  2011
4    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60  2011

In [25]:
# choose all columns for selected rows
df_sub_loc3 = df.loc[10:1000]
print(df_sub_loc3.head(2))
print(df_sub_loc3.tail(2))

# get rows for all columns indexed from 10 to 1000 (inclusive)

    Id      EmployeeName                                          JobTitle  \
10  11     ARTHUR KENNEY  ASSISTANT CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)   
11  12  PATRICIA JACKSON                   CAPTAIN III (POLICE DEPARTMENT)   

      BasePay  OvertimePay  OtherPay  Benefits   TotalPay  TotalPayBenefits  \
10  194999.39     71344.88   33149.9       NaN  299494.17         299494.17   
11   99722.00     87082.62  110804.3       NaN  297608.92         297608.92   

    Year  Notes         Agency  Status  
10  2011    NaN  San Francisco     NaN  
11  2011    NaN  San Francisco     NaN  
        Id  EmployeeName                               JobTitle    BasePay  \
999   1000    JOHN FAZIO              CLINICAL NURSE SPECIALIST  169572.00   
1000  1001  ROBERT KUZMA  CAPTAIN, EMERGENCYCY MEDICAL SERVICES  140546.84   

      OvertimePay  OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
999          0.00    250.00       NaN  169822.00         169822.00  2011   
1000      8577.69

### use .iloc[ , ] to select rows and columns

In [26]:
# choose 1 column
df_sub_iloc1 = df.iloc[100:120, 3:4]
# or use df_sub_iloc1 = df.iloc[100:120, 3]
df_sub_iloc1

# get rows indexed from 100-119 of the column index = 3

BasePay
100  168692.65
101  123105.00
102  105934.69
103   88400.70
104  105934.64
105  110474.93
106  149362.86
107  188708.00
108  188708.00
109  105934.67
110  206057.69
111   63811.18
112  220115.22
113  188708.00
114   88400.70
115  121628.34
116  219804.34
117  191054.19
118  150432.85
119  179648.00

In [27]:
# choose all rows for multiple columns
df_sub_iloc2 = df.iloc[:, 2:5]
df_sub_iloc2.head()

# get all rows for column index = 2, 3, 4

JobTitle    BasePay  OvertimePay
0  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18         0.00
1                 CAPTAIN III (POLICE DEPARTMENT)  155966.02    245131.88
2                 CAPTAIN III (POLICE DEPARTMENT)  212739.13    106088.18
3            WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00     56120.71
4    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60      9737.00

In [28]:
# choose all columns for selected rows
df_sub_iloc3 = df.iloc[10:1000]
print(df_sub_iloc3.head(2))
print(df_sub_iloc3.tail(2))

# get all rows with index from 100-999 for all columns

    Id      EmployeeName                                          JobTitle  \
10  11     ARTHUR KENNEY  ASSISTANT CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)   
11  12  PATRICIA JACKSON                   CAPTAIN III (POLICE DEPARTMENT)   

      BasePay  OvertimePay  OtherPay  Benefits   TotalPay  TotalPayBenefits  \
10  194999.39     71344.88   33149.9       NaN  299494.17         299494.17   
11   99722.00     87082.62  110804.3       NaN  297608.92         297608.92   

    Year  Notes         Agency  Status  
10  2011    NaN  San Francisco     NaN  
11  2011    NaN  San Francisco     NaN  
       Id EmployeeName                   JobTitle    BasePay  OvertimePay  \
998   999  ANNE HUGHES  CLINICAL NURSE SPECIALIST  169572.01          0.0   
999  1000   JOHN FAZIO  CLINICAL NURSE SPECIALIST  169572.00          0.0   

     OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  Notes  \
998     250.0       NaN  169822.01         169822.01  2011    NaN   
999     250.0       NaN  169822.00

## 5. Conditions_ selection criteria/ find specific values

### List comprehension

In [29]:
# list the columns with names that are at least 8 characters long
[col for col in df.columns if len(col) >=8]

['EmployeeName',
 'JobTitle',
 'OvertimePay',
 'OtherPay',
 'Benefits',
 'TotalPay',
 'TotalPayBenefits']

### == or !=, >, <= etc.

In [30]:
# find the NATHANIEL FORD' job title, basepay and other pay
df[df['EmployeeName'] == 'NATHANIEL FORD'][['JobTitle','BasePay','OtherPay']]

JobTitle    BasePay   OtherPay
0  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18  400184.25

In [31]:
# find the person with maximum basepay
df[df['BasePay']== df['BasePay'].max()]

# df.loc[df['BasePay'].idxmax()] # ot use this .idxmax()

Id    EmployeeName         JobTitle    BasePay  OvertimePay  \
72925  72926  Gregory P Suhr  Chief of Police  319275.01          0.0   

       OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  Notes  \
72925  20007.06  86533.21  339282.07         425815.28  2013    NaN   

              Agency  Status  
72925  San Francisco     NaN

In [32]:
# find the person with minumum basepay
df[df['BasePay']== df['BasePay'].min()]

# df[df['BasePay']== df['BasePay'].min()]['EmployeeName'] # find employee name

Id    EmployeeName      JobTitle  BasePay  OvertimePay  OtherPay  \
72832  72833  Irwin Sidharta  Junior Clerk  -166.01       249.02       0.0   

       Benefits  TotalPay  TotalPayBenefits  Year  Notes         Agency  \
72832      6.56     83.01             89.57  2012    NaN  San Francisco   

       Status  
72832     NaN

In [33]:
# find the average pay
df['BasePay'].mean()

# df.mean()['BasePay'] # or use this one

66325.44884050643

In [34]:
# find the avg pay by year
df.groupby('Year').mean()['BasePay']

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64